### DISCUSSION TITLE
"Named" locks

### DISCUSSION THREAD

**Discussion by davidbrochart** [agronholm/anyio#723] [Q&A]
It might be out of AnyIO's scope, but I often need a lock that is tied to an ID (or more generally a hashable object). For instance, resources on a file system could be represented by their path, and if I want to prevent concurrent access to the same resource, I'd like to write something like:
```py
resource_lock = ResourceLock()
async with resource_lock("path/to/file"):
    ...
```
Maybe I'm missing something, but nothing in [the synchronization primitives](https://anyio.readthedocs.io/en/stable/synchronization.html) seems to offer that functionality easily. Here is what I currently use:
```py
from typing import Any, Dict

from anyio import Event


class ResourceLock:
    _locks: Dict[Any, Event]

    def __init__(self):
        self._locks = {}

    def __call__(self, idx: Any):
        return _ResourceLock(idx, self._locks)


class _ResourceLock:
    _idx: Any
    _locks: Dict[Any, Event]
    _lock: Event

    def __init__(self, idx: Any, locks: Dict[Any, Event]):
        self._idx = idx
        self._locks = locks

    async def __aenter__(self):
        while True:
            if self._idx in self._locks:
                await self._locks[self._idx].wait()
            else:
                break
        self._locks[self._idx] = self._lock = Event()

    async def __aexit__(self, exc_type, exc_value, exc_tb):
        self._lock.set()
        del self._locks[self._idx]
```
It actually doesn't use any `Lock`. Do you see something wrong with that approach?  Or do you think it could be better implemented?

**Comment by agronholm** [comment]
I think this is indeed beyond AnyIO's scope, as it doesn't seem generic enough to warrant inclusion in this project. I also take issue with the fact that if the lock is contested by multiple tasks, it wakes up every single one of them when a lock is exited even when only one task just needs to be woken up. That `while` loop should not be necessary. On top of that, it doesn't guarantee FIFO ordering for lock acquisition.

**Reply by davidbrochart** [reply]
Right, actually it's better to use a `Lock` instead of an `Event`:
```py
from typing import Any

from anyio import Lock


class ResourceLock:
    _locks: dict[Any, Lock]

    def __init__(self):
        self._locks = {}

    def __call__(self, idx: Any):
        return _ResourceLock(idx, self._locks)


class _ResourceLock:
    _idx: Any
    _locks: dict[Any, Lock]
    _lock: Lock

    def __init__(self, idx: Any, locks: dict[Any, Lock]):
        self._idx = idx
        self._locks = locks

    async def __aenter__(self):
        if self._idx not in self._locks:
            self._locks[self._idx] = Lock()
        self._lock = self._locks[self._idx]
        await self._lock.acquire()

    async def __aexit__(self, exc_type, exc_value, exc_tb):
        self._lock.release()
        if self._idx in self._locks:
            del self._locks[self._idx]
```

**Comment by TayfurYldz** [comment]
The maintainer's concerns are real in the Event-based implementation: set() wakes every waiter, then all of them race through the while loop, and the dictionary update is not expressing FIFO ownership.

A keyed-lock helper can avoid both problems by storing a real Lock per key plus a reference/waiter count. The subtle part is cleanup: you must not remove the key merely because the current holder exits if another task has already obtained a reference to that same Lock and is waiting on it. Otherwise a third caller can create a new Lock for the same key and enter concurrently with the waiter.

Conceptually, protect the mapping itself with a short-lived guard lock:

1. under the guard, get/create the per-key Lock and increment its user count;
2. release the guard, then acquire the per-key Lock;
3. on exit, release the per-key Lock;
4. under the guard, decrement the count and remove the entry only when the count reaches zero.

AnyIO's Lock then supplies the single-waiter wake-up behavior instead of Event's broadcast. Cancellation while waiting also needs a finally path that decrements the reference count, otherwise abandoned keys leak.

If you only have a bounded/stable set of resource IDs, an even simpler solution is to keep a Lock for each resource for the lifetime of the owning object and skip cleanup entirely.

I agree with the maintainer that this is application-level keyed resource management rather than a core AnyIO primitive, but if you implement it locally I would use real per-key Locks with guarded lifetime accounting rather than the Event/while-loop pattern.
